staging/src/k8s.io/kubectl/pkg/kuberc/kuberc.go

```diff
@@ -0,0 +1,471 @@
+		if commandIndex+1 >= len(args) {
+			// command is the last item, we simply append just like appendArgs
+			args = append(args, aliasArgs.prependArgs...)
+		} else {
+			args = append(args[:commandIndex+1], append(aliasArgs.prependArgs, args[commandIndex+1:]...)...)
+		}
+	}
+	if len(aliasArgs.appendArgs) > 0 {
+		// appendArgs defined in kuberc should be appended to actual args.
+		args = append(args, aliasArgs.appendArgs...)
+	}
+	// Cobra (command.go#L1078) appends only root command's args into the actual args and ignores the others.
+	// We are appending the additional args defined in kuberc in here and
+	// expect that it will be passed along to the actual command.
+	rootCmd.SetArgs(args[1:])
+	return args, nil
+}
+
+// DefaultGetPreferences returns KubeRCConfiguration.
+// If users sets kuberc file explicitly in --kuberc flag, it has the highest
+// priority. If not specified, it looks for in KUBERC environment variable.
+// If KUBERC is also not set, it falls back to default .kuberc file at the same location
+// where kubeconfig's defaults are residing in.
+func DefaultGetPreferences(kuberc string) (*config.Preference, error) {
+	kubeRCFile := RecommendedKubeRCFile
+	explicitly := false
+	if kuberc != "" {
+		kubeRCFile = kuberc
+		explicitly = true
+	}
+
+	if kubeRCFile == "" && os.Getenv("KUBERC") != "" {
+		kubeRCFile = os.Getenv("KUBERC")
+		explicitly = true
+	}
+
+	kubeRCBytes, err := os.ReadFile(kubeRCFile)
+	if err != nil {
+		if !explicitly {
+			// We don't log if the kuberc file does not exist. Because user simply does not
+			// specify neither default kuberc file nor explicitly pass it.
+			// We'll continue to default behavior without raising any error.
+			if !os.IsNotExist(err) {
+				klog.V(4).Infof("error reading kuberc file %q: %v", kubeRCFile, err)
+			}
+			return nil, nil
+		}
+		// Kuberc is specified by user via flag or env variable and it gets an error.
+		return nil, err
+	}
+
+	var pref config.Preference
+	_, gvk, err := strictCodecs.UniversalDecoder().Decode(kubeRCBytes, nil, &pref)
+	if err != nil {
+		if explicitly {
+			// explicitly specified kuberc can't be decoded and we short cut the process
+			return nil, fmt.Errorf("could not be decoded gvk %s, err: %w", gvk, err)
+		}
+		// default kuberc is incompatible with this version, or it simply is invalid.
+		// falling back to lenient decoding to do our best.
+		_, gvk2, err2 := lenientCodecs.UniversalDecoder().Decode(kubeRCBytes, nil, &pref)
+		if err2 != nil {
+			return nil, fmt.Errorf("could not be decoded gvk %s, err: %w", gvk2, err2)
+		}
+		klog.Warningf("gvk %s could not be decoded with strict decoding, continue with less strict decoding: %s", gvk, err)
+		gvk = gvk2
+	}
+
+	expectedGK := schema.GroupKind{
+		Group: config.SchemeGroupVersion.Group,
+		Kind:  "Preference",
+	}
+	if gvk.GroupKind() != expectedGK {
+		return nil, fmt.Errorf("unsupported preference GVK %s", gvk.GroupKind().String())
+	}
```

Comment:
the fallback here is not quite right... when strict decoding fails, we should always be falling back to lenient decoding, while remembering the strict decoding hit an issue

1. there are errors (when even lenient decoding failed, or the group / kind wasn't correct) and warnings (strict decoding failed, lenient decoding succeeded)
2. for warnings, we can just warn; for errors, if kuberc wasn't explicitly requested, I still think we should just warn

Reply:
@liggitt I've updated the PR according to your suggestion. Let me know what you think when you have any cycles. Thank you.